sdk/core/Azure.Core.Experimental/src/Primitives/BinaryData.cs

```diff
@@ -0,0 +1,121 @@
+﻿// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+using System;
+using System.ComponentModel;
+using System.IO;
+using System.Linq;
+using System.Text;
+
+namespace Azure.Core.Experimental.Primitives
+{
+    /// <summary>
+    /// Abstraction for a payload of binary data.
+    /// </summary>
+    public class BinaryData
+    {
+        /// <summary>
+        /// The backing store for the <see cref="BinaryData"/> instance.
+        /// </summary>
+        internal ReadOnlyMemory<byte> Memory { get; }
+
+        /// <summary>
+        /// Creates a binary data instance from bytes.
+        /// </summary>
+        /// <param name="data">Byte data.</param>
+        public BinaryData(ReadOnlyMemory<byte> data)
+        {
+            Memory = data;
+        }
+
+        /// <summary>
+        /// Creates a binary data instance from a string
+        /// using UTF-8 encoding.
+        /// </summary>
+        /// <param name="data">The string data.</param>
+        /// <returns>A <see cref="BinaryData"/> instance.</returns>
+        public static BinaryData Create(string data) =>
+            Create(data, Encoding.UTF8);
+
+        /// <summary>
+        /// Creates a binary data instance from a string
+        /// using the specified encoding.
+        /// </summary>
+        /// <param name="data">The string data.</param>
+        /// <param name="encoding">The encoding to use when converting
+        /// the data to bytes.</param>
+        /// <returns>A <see cref="BinaryData"/> instance.</returns>
+        public static BinaryData Create(string data, Encoding encoding)
+        {
+            Argument.AssertNotNull(encoding, nameof(encoding));
+            return new BinaryData(encoding.GetBytes(data));
+        }
+
+        /// <summary>
+        /// Creates a binary data instance from the specified stream.
+        /// </summary>
+        /// <param name="stream">Stream containing the data.</param>
+        /// <returns>A <see cref="BinaryData"/> instance.</returns>
+        public static BinaryData Create(Stream stream)
+        {
+            Argument.AssertNotNull(stream, nameof(stream));
```

Comment:
Do we also want to force a fixed size on this?  Streams can be longer than int.MaxValue, for example.  I'd start by enforcing it's at least that size or less if the stream has a length.

Reply:
oops missed this comment..